README.md
# utf8-buffer
ES6 module to encode and decode UTF-8 strings.  
Copyright (c) 2018-2019 Rafael da Silva Rocha.  
https://github.com/rochars/utf8-buffer

[![NPM version](https://img.shields.io/npm/v/utf8-buffer.svg?style=for-the-badge)](https://www.npmjs.com/package/utf8-buffer) [![Docs](https://img.shields.io/badge/docs-online-blue.svg?style=for-the-badge)](https://rochars.github.io/utf8-buffer/docs/index.html) [![Tests](https://img.shields.io/badge/tests-online-blue.svg?style=for-the-badge)](https://rochars.github.io/utf8-buffer/test/dist/browser.html)  
[![Codecov](https://img.shields.io/codecov/c/github/rochars/utf8-buffer.svg?style=flat-square)](https://codecov.io/gh/rochars/utf8-buffer) [![Unix Build](https://img.shields.io/travis/rochars/utf8-buffer.svg?style=flat-square)](https://travis-ci.org/rochars/utf8-buffer) [![Windows Build](https://img.shields.io/appveyor/ci/rochars/utf8-buffer.svg?style=flat-square&logo=appveyor)](https://ci.appveyor.com/project/rochars/utf8-buffer) [![Scrutinizer](https://img.shields.io/scrutinizer/g/rochars/utf8-buffer.svg?style=flat-square&logo=scrutinizer)](https://scrutinizer-ci.com/g/rochars/utf8-buffer/)

**utf8-buffer** is a ES6 module to encode and decode UTF-8 strings.

## Install
```
npm install utf8-buffer
```

## Use

### Browser
Use the **utf8-buffer.js** file in the */dist* folder:
```html
<script src="./dist/utf8-buffer.js"></script>
```

Or load it from the [jsDelivr](https://cdn.jsdelivr.net/npm/utf8-buffer) CDN:
```html
<script src="https://cdn.jsdelivr.net/npm/utf8-buffer"></script>
```

Or load it from [unpkg](https://unpkg.com/utf8-buffer):
```html
<script src="https://unpkg.com/utf8-buffer"></script>
```

### Node
```javascript
const utf8Buffer = require('utf8-buffer');
```

Or **import** just what you need:
```javascript
import {pack, unpack} from 'utf8-buffer';
```

## About
Only UTF-8 strings with a max of 4 bytes per character are supported. **BOM** is kept untouched. Invalid characters are replaced with *Unicode Character 'REPLACEMENT CHARACTER' (U+FFFD)*.

## API
```javascript
/**
 * Read a string of UTF-8 characters from a byte buffer.
 * Invalid characters are replaced with 'REPLACEMENT CHARACTER' (U+FFFD).
 * @see https://encoding.spec.whatwg.org/#the-encoding
 * @see https://stackoverflow.com/a/34926911
 * @param {!Uint8Array|!Array<number>} buffer A byte buffer.
 * @param {number=} start The buffer index to start reading.
 * @param {?number=} end The buffer index to stop reading.
 *   Assumes the buffer length if undefined.
 * @return {string}
 */
export function unpack(buffer, start=0, end=buffer.length) {}

/**
 * Write a string of UTF-8 characters to a byte buffer.
 * @see https://encoding.spec.whatwg.org/#utf-8-encoder
 * @param {string} str The string to pack.
 * @param {!Uint8Array|!Array<number>} buffer The buffer to pack the string to.
 * @param {number=} index The buffer index to start writing.
 * @return {number} The next index to write in the buffer.
 */
export function pack(str, buffer, index=0) {}
```

## Contributing
**utf8-buffer** welcomes all contributions from anyone willing to work in good faith with other contributors and the community. No contribution is too small and all contributions are valued.

See [CONTRIBUTING.md](https://github.com/rochars/utf8-buffer/blob/master/CONTRIBUTING.md) for details.

### Style guide
**utf8-buffer** code should follow the Google JavaScript Style Guide:  
https://google.github.io/styleguide/jsguide.html

### Code of conduct
This project is bound by a code of conduct: The [Contributor Covenant, version 1.4](https://github.com/rochars/utf8-buffer/blob/master/CODE_OF_CONDUCT.md), also available at https://www.contributor-covenant.org/version/1/4/code-of-conduct.html

## LICENSE
Copyright (c) 2018-2019 Rafael da Silva Rocha.

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# CHANGELOG

## 1.0.0 - 2019-12-31
- New package structure:
	* dist file is "./dist/utf8-buffer.js", a UMD served as "main"
	* ES6 source is "./index.js", served as "module"

## 0.2.0 (2018-08-05)
- unpack(buffer, start, end) API change: end now is non-inclusive

## 0.1.0 (2018-07-27)
- pack(str):Uint8Array updated to pack(str, buffer, index=0):number
	- *buffer* is the buffer you are writing to (ideally Uint8Array, but will work the same with Array or any Array-like object).
	- *index* is a optional param with the buffer index to start writing. Assumes zero by default.
	- The return value is a *number*, the next index to write in the array.
- unpack(buffer, index, len) updated to unpack(buffer, start, end)
	- *start* is the buffer index to start reading.
	- *end* is the buffer index to end reading.
- works in IE8+.


test/loader.js
/**
 * Copyright (c) 2018-2019 Rafael da Silva Rocha.
 * https://github.com/rochars/utf8-buffer
 *
 */

let utf8Buffer;

// UMD bundle
if (process.argv[3] == '--umd') {
	console.log('umd tests');
	utf8Buffer = require('../dist/utf8-buffer.js');

// Source
} else {
	require = require("esm")(module);
	global.module = module;
	console.log('Source tests');
	utf8Buffer = require('../index.js');
}

module.exports = utf8Buffer;


test/src/UTF-8.js
/*
 * Copyright (c) 2018 Rafael da Silva Rocha.
 */

/**
 * @fileoverview UTF-8 tests.
 * @see https://github.com/rochars/utf8-buffer
 */

var utf8Buffer = utf8Buffer || require('../../test/loader.js');
var assert = assert || require('assert');
var Buffer = Buffer || false;

describe('replace invalid characters', function() {
    it('returns the next index to write', function() {
        buffer = [];
        let index = utf8Buffer.pack('ab', buffer);
        utf8Buffer.pack('cd', buffer, index);
        assert.deepEqual(buffer, [97,98,99,100]);
    });
    it('writes from zero if no index is given', function() {
        buffer = [];
        let index = utf8Buffer.pack('ab', buffer);
        utf8Buffer.pack('cd', buffer);
        assert.deepEqual(buffer, [99,100]);
    });
    it('replaces invalid 2 byte UTF-8 char (2nd byte)', function() {
        assert.equal(utf8Buffer.unpack([0xa0,0xa1]), '\uFFFD\uFFFD');
    });
    // 3 bytes
    it('replaces invalid 3 byte UTF-8 char (2rd byte)', function() {
        assert.equal(utf8Buffer.unpack([0xe2,0x28,0xa1]), '\uFFFD');
    });
    it('replaces invalid 3 byte UTF-8 char (2rd byte)', function() {
        assert.equal(utf8Buffer.unpack(
            [97,0xe2,0x28,0xa1,98]), 'a\uFFFDb');
    });
    it('replaces invalid 3 byte UTF-8 char (3rd byte)', function() {
        assert.equal(utf8Buffer.unpack([0xe2,0x82,0x28]), '\uFFFD');
    });
    it('replaces invalid 3 byte UTF-8 char (3rd byte)', function() {
        assert.equal(utf8Buffer.unpack([0xe2,0xE0,0x28]), '\uFFFD');
    });
    it('replaces invalid 3 byte UTF-8 char (3rd byte)', function() {
        assert.equal(utf8Buffer.unpack([0xe2,0xED,0x28]), '\uFFFD');
    });
    // 4 bytes
    it('replaces invalid 4 byte UTF-8 char (2th byte)', function() {
        assert.equal(utf8Buffer.unpack([0xf0,0x28,0x8c,0xbc]), '\uFFFD');
    });
    it('replaces invalid 4 byte UTF-8 char (3th byte)', function() {
        assert.equal(utf8Buffer.unpack([0xf0,0x90,0x28,0xbc]), '\uFFFD');
    });
    it('replaces invalid 4 byte UTF-8 char (4th byte)', function() {
        assert.equal(utf8Buffer.unpack(
            [97,0xf0,0x28,0x8c,0x28,240,175,167,159,240,175,167,159]),
        'a\uFFFD輸輸');
    });
    it('replaces invalid 4 byte UTF-8 char (4th byte)', function() {
        assert.equal(utf8Buffer.unpack([0xf0,0xF4,0x8c,0x28]), '\uFFFD');
    });
    it('replaces invalid 4 byte UTF-8 char (4th byte)', function() {
        assert.equal(utf8Buffer.unpack([0xf0,0xF0,0x8c,0x28]), '\uFFFD');
    });

})

// pack
describe('unpack UTF-8 strings, 4 bytes', function() {
    it('pack BOM + 輸 as a byte array', function() {
        // 輸 CJK COMPATIBILITY IDEOGRAPH-2F9DF (U+2F9DF) f0 af a7 9f
        var chars = '\ufeff輸';
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
        assert.deepEqual(
            [
                0xEF,0xBB,0xBF,
                240,175,167,159],
            buffer);
    });
});
describe('unpack UTF-8 strings, 4 bytes', function() {
    it('pack 輸 + BOM + 輸 as a byte array', function() {
        // 輸   CJK COMPATIBILITY IDEOGRAPH-2F9DF (U+2F9DF) f0 af a7 9f
        var chars = '輸\ufeff輸';
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
        assert.deepEqual(
            [
                240,175,167,159,
                0xEF,0xBB,0xBF,
                240,175,167,159],
            buffer);
    });
});
describe('pack UTF-8 strings, 4 bytes', function() {
    it('pack 輸 as a byte array', function() {
        // 輸 CJK COMPATIBILITY IDEOGRAPH-2F9DF (U+2F9DF) f0 af a7 9f
        var chars = '輸';
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
        assert.deepEqual(
            [240,175,167,159],
            buffer);
    });
    it('pack 輸輸 as a byte array', function() {
        var chars = '輸輸';
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
        assert.deepEqual(
            [240,175,167,159,
                240,175,167,159],
            buffer);
    });
    it('pack 輸輸笠߹~$ as a byte array', function() {
        var chars = '輸輸笠߹~$';
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
        assert.deepEqual(
            [
                240,175,167,159,
                240,175,167,159,
                239, 167, 184,
                223, 185,
                126,
                36],
            buffer);
    });
    it('pack 輸輸笠߹~$輸 as a byte array', function() {
        var chars = '輸輸笠߹~$輸';
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
        assert.deepEqual(
            [
                240,175,167,159,
                240,175,167,159,
                239, 167, 184,
                223, 185,
                126,
                36,
                240,175,167,159],
            buffer);
    });
});
describe('pack UTF-8 strings, 3 bytes', function() {
    it('pack 笠 as a byte array', function() {
        // CJK COMPATIBILITY IDEOGRAPH-F9F8 (U+F9F8) ef a7 b8
        var chars = '笠';
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
        assert.deepEqual(
            [239, 167, 184],
            buffer);
    });
    it('pack 笠笠 as a byte array', function() {
        var chars = '笠笠';
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
        assert.deepEqual(
            [
                239, 167, 184,
                239, 167, 184],
            buffer);
    });
    it('pack 笠笠߹~$ as a byte array', function() {
        var chars = '笠笠߹~$';
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
        assert.deepEqual(
            [
                239, 167, 184,
                239, 167, 184,
                223, 185,
                126,
                36],
            buffer);
    });
    it('pack 笠笠߹~$笠 as a byte array', function() {
        var chars = '笠笠߹~$笠';
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
        assert.deepEqual(
            [
                239, 167, 184,
                239, 167, 184,
                223, 185,
                126,
                36,
                239, 167, 184],
            buffer);
    });
});
describe('pack UTF-8 strings, 2 bytes', function() {
    it('pack ߹ as a byte array', function() {
        var chars = '߹'; // NKO EXCLAMATION MARK (U+07F9)
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
        assert.deepEqual(
            [223, 185],
            buffer);
    });
    it('pack ߹߹ as a byte array', function() {
        var chars = '߹߹';
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
        assert.deepEqual(
            [223, 185, 223, 185],
            buffer);
    });
    it('pack ߹߹~$ as a byte array', function() {
        var chars = '߹߹~$';
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
        assert.deepEqual(
            [223, 185, 223, 185, 126, 36],
            buffer);
    });
    it('pack ߹߹~$߹ as a byte array', function() {
        var chars = '߹߹~$߹';
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
        assert.deepEqual(
            [223, 185, 223, 185, 126, 36, 223, 185],
            buffer);
    });
});
describe('pack UTF-8 strings, 1 byte', function() {
    it('pack ~ as a byte array', function() {
        var chars = '~'; // TILDE (U+007E)  7e
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
        assert.deepEqual(
            [126],
            buffer);
    });
    it('pack ~~ as a byte array', function() {
        var chars = '~~';
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
       assert.deepEqual(
            [126, 126],
            buffer);
    });
    it('pack ~~$ as a byte array', function() {
        var chars = '~~$';
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
        assert.deepEqual(
            [126, 126, 36],
            buffer);
    });
    it('pack ~~$~ as a byte array', function() {
        var chars = '~~$~';
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
        assert.deepEqual(
            [126, 126, 36, 126],
            buffer);
    });
});
describe('pack ASCII strings', function() {
    it('pack $ as a byte array', function() {
        var chars = '$';
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
        assert.deepEqual([36], buffer);
    });
    it('pack $$ as a byte array', function() {
        var chars = '$$';
        var buffer = [];
        utf8Buffer.pack(chars, buffer);
        assert.deepEqual([36, 36], buffer);
    });
});

// unpack
describe('unpack UTF-8 strings, 4 bytes with BOM', function() {
    it('unpack BOM + 輸 from a byte buffer', function() {
        // 輸 CJK COMPATIBILITY IDEOGRAPH-2F9DF (U+2F9DF) f0 af a7 9f
        var chars = [
            0xEF,0xBB,0xBF,
            240,175,167,159];
        assert.equal('\ufeff輸', utf8Buffer.unpack(chars));
    });
});
describe('unpack UTF-8 strings', function() {
    it('unpack 𒈓 from a byte buffer', function() {
        // CUNEIFORM SIGN LU2 OPPOSING LU2 (U+12213)
        var chars = [0xF0, 0x92, 0x88, 0x93];
        assert.equal('𒈓', utf8Buffer.unpack(chars));
    });
    it('unpack 𒈓𒈓 from a byte buffer', function() {
        // CUNEIFORM SIGN LU2 OPPOSING LU2 (U+12213)
        var chars = [
            0xF0, 0x92, 0x88, 0x93,
            0xF0, 0x92, 0x88, 0x93];
        assert.equal('𒈓𒈓', utf8Buffer.unpack(chars));
    });
    it('unpack 𒈓𒈓$ from a byte buffer', function() {
        // CUNEIFORM SIGN LU2 OPPOSING LU2 (U+12213)
        var chars = [
            0xF0, 0x92, 0x88, 0x93,
            0xF0, 0x92, 0x88, 0x93,
            36];
        assert.equal('𒈓𒈓$', utf8Buffer.unpack(chars));
    });
    it('unpack 𒈓𒈓$笠 from a byte buffer', function() {
        // CUNEIFORM SIGN LU2 OPPOSING LU2 (U+12213)
        var chars = [
            0xF0, 0x92, 0x88, 0x93,
            0xF0, 0x92, 0x88, 0x93,
            36,
            239, 167, 184];
        assert.equal('𒈓𒈓$笠', utf8Buffer.unpack(chars));
    });
    it('unpack BOM + 𒈓𒈓$笠 from a byte buffer', function() {
        // CUNEIFORM SIGN LU2 OPPOSING LU2 (U+12213)
        var chars = [
            0xEF,0xBB,0xBF,
            0xF0, 0x92, 0x88, 0x93,
            0xF0, 0x92, 0x88, 0x93,
            36,
            239, 167, 184];
        assert.equal('\ufeff𒈓𒈓$笠', utf8Buffer.unpack(chars));
    });
});
describe('unpack UTF-8 strings, 4 bytes', function() {
    it('unpack 輸 from a byte buffer', function() {
        // 輸   CJK COMPATIBILITY IDEOGRAPH-2F9DF (U+2F9DF) f0 af a7 9f
        var chars = [240,175,167,159];
        assert.equal('輸', utf8Buffer.unpack(chars));
    });
    it('unpack 輸輸 from a byte buffer', function() {
        var chars = [
            240,175,167,159,
            240,175,167,159];
        assert.equal('輸輸', utf8Buffer.unpack(chars));
    });
    it('unpack 輸輸笠߹~$ from a byte buffer', function() {
        var chars = [
            240,175,167,159,
            240,175,167,159,
            239, 167, 184,
            223, 185,
            126,
            36];
        assert.equal('輸輸笠߹~$', utf8Buffer.unpack(chars));
    });
    it('unpack 輸輸笠߹~$輸 from a byte buffer', function() {
        var chars = [
            240,175,167,159,
            240,175,167,159,
            239, 167, 184,
            223, 185,
            126,
            36,
            240,175,167,159];
        assert.equal('輸輸笠߹~$輸', utf8Buffer.unpack(chars));
    });
});
describe('unpack UTF-8 strings, 3 bytes', function() {
    it('unpack 笠 from a byte buffer', function() {
        // CJK COMPATIBILITY IDEOGRAPH-F9F8 (U+F9F8) ef a7 b8 // 239 167 184
        var chars = [239, 167, 184];
        assert.equal('笠', utf8Buffer.unpack(chars));
    });
    it('unpack 笠笠 from a byte buffer', function() {
        var chars = [239, 167, 184, 239, 167, 184];
        assert.equal('笠笠', utf8Buffer.unpack(chars));
    });
    it('unpack 笠笠߹~$ from a byte buffer', function() {
        var chars = [
            239, 167, 184,
            239, 167, 184,
            223, 185,
            126,
            36];
        assert.equal('笠笠߹~$', utf8Buffer.unpack(chars));
    });
    it('unpack 笠笠߹~$笠 from a byte buffer', function() {
        var chars = [
            239, 167, 184,
            239, 167, 184,
            223, 185,
            126,
            36,
            239, 167, 184];
        assert.equal('笠笠߹~$笠',utf8Buffer.unpack(chars));
    });
});
describe('unpack UTF-8 strings, 2 bytes', function() {
    it('unpack å from a byte buffer', function() {
        var chars = [195,165];
        assert.equal('å', utf8Buffer.unpack(chars));
    });
    it('unpack ߹ from a byte buffer', function() {
        var chars = [223, 185]; // NKO EXCLAMATION MARK (U+07F9)
        assert.equal('߹', utf8Buffer.unpack(chars));
    });
    it('unpack ߹߹ from a byte buffer', function() {
        var chars = [223, 185, 223, 185];
        assert.equal('߹߹', utf8Buffer.unpack(chars));
    });
    it('unpack ߹߹~$ from a byte buffer', function() {
        var chars = [223, 185, 223, 185, 126, 36];
        assert.equal('߹߹~$', utf8Buffer.unpack(chars));
    });
    it('unpack ߹߹~$߹ from a byte buffer', function() {
        var chars = [223, 185, 223, 185, 126, 36, 223, 185];
        assert.equal('߹߹~$߹', utf8Buffer.unpack(chars));
    });
});
describe('unpack UTF-8 strings, 1 byte', function() {
    it('unpack ~ from a byte buffer', function() {
        var chars = [126]; // TILDE (U+007E)  7e
        assert.equal('~', utf8Buffer.unpack(chars));
    });
    it('unpack ~~ from a byte buffer', function() {
        var chars = [126, 126];
        assert.equal('~~', utf8Buffer.unpack(chars));
    });
    it('unpack ~~$ from a byte buffer', function() {
        var chars = [126, 126, 36];
        assert.equal('~~$', utf8Buffer.unpack(chars));
    });
    it('unpack ~~$~ from a byte buffer', function() {
        var chars = [126, 126, 36, 126];
        assert.equal('~~$~', utf8Buffer.unpack(chars));
    });
});
describe('unpack ASCII strings', function() {
    it('unpack $ from a byte buffer', function() {
        var chars = [36];
        assert.equal('$', utf8Buffer.unpack(chars));
    });
    it('unpack $$ from a byte buffer', function() {
        var chars = [36, 36];
        assert.equal('$$', utf8Buffer.unpack(chars));
    });
});


index.js
/*
 * Copyright (c) 2018 Rafael da Silva Rocha.
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 */

/**
 * @fileoverview Functions to serialize and deserialize UTF-8 strings.
 * @see https://github.com/rochars/utf8-buffer
 * @see https://encoding.spec.whatwg.org/#the-encoding
 * @see https://encoding.spec.whatwg.org/#utf-8-encoder
 */

/** @module utf8-buffer */

/**
 * Read a string of UTF-8 characters from a byte buffer.
 * Invalid characters are replaced with 'REPLACEMENT CHARACTER' (U+FFFD).
 * @see https://encoding.spec.whatwg.org/#the-encoding
 * @see https://stackoverflow.com/a/34926911
 * @param {!Uint8Array|!Array<number>} buffer A byte buffer.
 * @param {number=} start The buffer index to start reading.
 * @param {?number=} end The buffer index to stop reading.
 *   Assumes the buffer length if undefined.
 * @return {string}
 */
export function unpack(buffer, start=0, end=buffer.length) {
  /** @type {string} */
  let str = '';
  for(let index = start; index < end;) {
    /** @type {number} */
    let lowerBoundary = 0x80;
    /** @type {number} */
    let upperBoundary = 0xBF;
    /** @type {boolean} */
    let replace = false;
    /** @type {number} */
    let charCode = buffer[index++];
    if (charCode >= 0x00 && charCode <= 0x7F) {
      str += String.fromCharCode(charCode);
    } else {
      /** @type {number} */
      let count = 0;
      if (charCode >= 0xC2 && charCode <= 0xDF) {
        count = 1;
      } else if (charCode >= 0xE0 && charCode <= 0xEF ) {
        count = 2;
        if (buffer[index] === 0xE0) {
          lowerBoundary = 0xA0;
        }
        if (buffer[index] === 0xED) {
          upperBoundary = 0x9F;
        }
      } else if (charCode >= 0xF0 && charCode <= 0xF4 ) {
        count = 3;
        if (buffer[index] === 0xF0) {
          lowerBoundary = 0x90;
        }
        if (buffer[index] === 0xF4) {
          upperBoundary = 0x8F;
        }
      } else {
        replace = true;
      }
      charCode = charCode & (1 << (8 - count - 1)) - 1;
      for (let i = 0; i < count; i++) {
        if (buffer[index] < lowerBoundary || buffer[index] > upperBoundary) {
          replace = true;
        }
        charCode = (charCode << 6) | (buffer[index] & 0x3f);
        index++;
      }
      if (replace) {
        str += String.fromCharCode(0xFFFD);
      } 
      else if (charCode <= 0xffff) {
        str += String.fromCharCode(charCode);
      } else {
        charCode -= 0x10000;
        str += String.fromCharCode(
          ((charCode >> 10) & 0x3ff) + 0xd800,
          (charCode & 0x3ff) + 0xdc00);
      }
    }
  }
  return str;
}

/**
 * Write a string of UTF-8 characters to a byte buffer.
 * @see https://encoding.spec.whatwg.org/#utf-8-encoder
 * @param {string} str The string to pack.
 * @param {!Uint8Array|!Array<number>} buffer The buffer to pack the string to.
 * @param {number=} index The buffer index to start writing.
 * @return {number} The next index to write in the buffer.
 */
export function pack(str, buffer, index=0) {
  for (let i = 0, len = str.length; i < len; i++) {
    /** @type {number} */
    let codePoint = str.codePointAt(i);
    if (codePoint < 128) {
      buffer[index] = codePoint;
      index++;
    } else {
      /** @type {number} */
      let count = 0;
      /** @type {number} */
      let offset = 0;
      if (codePoint <= 0x07FF) {
        count = 1;
        offset = 0xC0;
      } else if(codePoint <= 0xFFFF) {
        count = 2;
        offset = 0xE0;
      } else if(codePoint <= 0x10FFFF) {
        count = 3;
        offset = 0xF0;
        i++;
      }
      buffer[index] = (codePoint >> (6 * count)) + offset;
      index++;
      while (count > 0) {
        buffer[index] = 0x80 | (codePoint >> (6 * (count - 1)) & 0x3F);
        index++;
        count--;
      }
    }
  }
  return index;
}


index.d.ts
// Type definitions for utf8-buffer 1.0.0
// Project: https://github.com/rochars/byte-data
// Definitions by: Rafael da Silva Rocha <https://github.com/rochars>
// Definitions: https://github.com/rochars/utf8-buffer

/**
 * Read a string of UTF-8 characters from a byte buffer.
 * Invalid characters are replaced with 'REPLACEMENT CHARACTER' (U+FFFD).
 * @see https://encoding.spec.whatwg.org/#the-encoding
 * @see https://stackoverflow.com/a/34926911
 * @param {!Uint8Array|!Array<number>} buffer A byte buffer.
 * @param {number=} start The buffer index to start reading.
 * @param {?number=} end The buffer index to stop reading.
 *   Assumes the buffer length if undefined.
 * @return {string}
 */
export function unpack(
	buffer: Uint8Array|number[],
	start: number,
	end?: number): string;

/**
 * Write a string of UTF-8 characters to a byte buffer.
 * @see https://encoding.spec.whatwg.org/#utf-8-encoder
 * @param {string} str The string to pack.
 * @param {!Uint8Array|!Array<number>} buffer The buffer to pack the string to.
 * @param {number=} index The buffer index to start writing.
 * @return {number} The next index to write in the buffer.
 */
export function pack(
	str: string,
	buffer: Uint8Array|number[],
	index?: number): number;


main.js
/*
 * Copyright (c) 2018 Rafael da Silva Rocha.
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 */

/**
 * @fileoverview Functions to serialize and deserialize UTF-8 strings.
 * @see https://github.com/rochars/utf8-buffer
 * @see https://encoding.spec.whatwg.org/#the-encoding
 * @see https://encoding.spec.whatwg.org/#utf-8-encoder
 */

/** @module utf8-buffer */

/**
 * Read a string of UTF-8 characters from a byte buffer.
 * Invalid characters are replaced with 'REPLACEMENT CHARACTER' (U+FFFD).
 * @see https://encoding.spec.whatwg.org/#the-encoding
 * @see https://stackoverflow.com/a/34926911
 * @param {!Uint8Array|!Array<number>} buffer A byte buffer.
 * @param {number=} start The buffer index to start reading.
 * @param {?number=} end The buffer index to stop reading.
 *   Assumes the buffer length if undefined.
 * @return {string}
 */
export function unpack(buffer, start=0, end=buffer.length) {
  /** @type {string} */
  let str = "";
  for(let index = start; index < end;) {
    /** @type {number} */
    let lowerBoundary = 0x80;
    /** @type {number} */
    let upperBoundary = 0xBF;
    /** @type {boolean} */
    let replace = false;
    /** @type {number} */
    let charCode = buffer[index++];
    if (charCode >= 0x00 && charCode <= 0x7F) {
      str += String.fromCharCode(charCode);
    } else {
      /** @type {number} */
      let count = 0;
      if (charCode >= 0xC2 && charCode <= 0xDF) {
        count = 1;
      } else if (charCode >= 0xE0 && charCode <= 0xEF ) {
        count = 2;
        if (buffer[index] === 0xE0) {
          lowerBoundary = 0xA0;
        }
        if (buffer[index] === 0xED) {
          upperBoundary = 0x9F;
        }
      } else if (charCode >= 0xF0 && charCode <= 0xF4 ) {
        count = 3;
        if (buffer[index] === 0xF0) {
          lowerBoundary = 0x90;
        }
        if (buffer[index] === 0xF4) {
          upperBoundary = 0x8F;
        }
      } else {
        replace = true;
      }
      charCode = charCode & (1 << (8 - count - 1)) - 1;
      for (let i = 0; i < count; i++) {
        if (buffer[index] < lowerBoundary || buffer[index] > upperBoundary) {
          replace = true;
        }
        charCode = (charCode << 6) | (buffer[index] & 0x3f);
        index++;
      }
      if (replace) {
        str += String.fromCharCode(0xFFFD);
      } 
      else if (charCode <= 0xffff) {
        str += String.fromCharCode(charCode);
      } else {
        charCode -= 0x10000;
        str += String.fromCharCode(
          ((charCode >> 10) & 0x3ff) + 0xd800,
          (charCode & 0x3ff) + 0xdc00);
      }
    }
  }
  return str;
}

/**
 * Write a string of UTF-8 characters to a byte buffer.
 * @see https://encoding.spec.whatwg.org/#utf-8-encoder
 * @param {string} str The string to pack.
 * @param {!Uint8Array|!Array<number>} buffer The buffer to pack the string to.
 * @param {number=} index The buffer index to start writing.
 * @return {number} The next index to write in the buffer.
 */
export function pack(str, buffer, index=0) {
  for (let i = 0, len = str.length; i < len; i++) {
    /** @type {number} */
    let codePoint = str.codePointAt(i);
    if (codePoint < 128) {
      buffer[index] = codePoint;
      index++;
    } else {
      /** @type {number} */
      let count = 0;
      /** @type {number} */
      let offset = 0;
      if (codePoint <= 0x07FF) {
        count = 1;
        offset = 0xC0;
      } else if(codePoint <= 0xFFFF) {
        count = 2;
        offset = 0xE0;
      } else if(codePoint <= 0x10FFFF) {
        count = 3;
        offset = 0xF0;
        i++;
      }
      buffer[index] = (codePoint >> (6 * count)) + offset;
      index++;
      while (count > 0) {
        buffer[index] = 0x80 | (codePoint >> (6 * (count - 1)) & 0x3F);
        index++;
        count--;
      }
    }
  }
  return index;
}
